### PR TITLE
Do not use fallocate in swap file creation on xfs.

### PR DESCRIPTION
**Pull request by otubo**
When creating a swap file on an xfs filesystem, fallocate cannot be used.
Doing so results in failure of swapon and a message like:
 swapon: swapfile has holes

The solution here is to maintain a list (currently containing only XFS)
of filesystems where fallocate cannot be used. The, on those fileystems
use the slower but functional 'dd' method.

Based on initial pull request:
https://code.launchpad.net/~adobrawy/cloud-init/+git/cloud-init/+merge/354680

And further comments from Scott Moser:
http://paste.ubuntu.com/p/8pVhczVqG3/

Signed-off-by: Eduardo Otubo <otubo@redhat.com>

### CODE REVIEW

**Comment by otubo** [comment]
This is just the first draft RFC and check. No tests written yet.

File: cloudinit/config/cc_mounts.py [hunk 1]
```diff
@@ -245,20 +315,8 @@ def setup_swapfile(fname, size=None, maxsize=None):
...
-                      args=[['sh', '-c',
-                             ('rm -f "$1" && umask 0066 && '
-                              '{ fallocate -l "${2}M" "$1" || '
-                              'dd if=/dev/zero "of=$1" bs=1M "count=$2"; } && '
-                              'mkswap "$1" || { r=$?; rm -f "$1"; exit $r; }'),
-                             'setup_swap', fname, mbsize]])
-
-    except Exception as e:
-        raise IOError("Failed %s: %s" % (msg, e))
+    util.log_time(LOG.debug, msg="Setting up swap file", func=setup_swapfile,
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: Should this be:

```suggestion
    util.log_time(LOG.debug, msg="Setting up swap file", func=create_swapfile,
```

?

File: cloudinit/config/cc_mounts.py [hunk 2]
```diff
@@ -223,6 +223,76 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
 
 
+def get_fstype(path):
+    """Return the type of the filesystem that path is on.
+    If path is a mount point itself, then the type of the filesystem mounted
+    on path rather than its parent dir."""
+    return util.get_mount_info(path)[1]
+
+
+def create_swapfile(fname, size, method="auto"):
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: AFAICT, this is never called with `method`<sup>0</sup>. Should we just drop the `method` parameter and simplify the function implementation?

<sup>0</sup> Well, at the moment, it's never called at all. See my comment below. ;-)

File: cloudinit/config/cc_mounts.py [hunk 3]
```diff
@@ -223,6 +223,76 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+
+    def unlink_f(fname):
+        if os.path.exists(fname):
+            os.unlink(fname)
+
+    unlink_f(fname)
+
+    msg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
+    if method == fallocate:
+        cmd = ['fallocate', '-l', '%dM' % size, fname]
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: The bodies of this if statement and the one that follows are almost identical. Should we be refactoring that duplication out to a helper?

File: cloudinit/config/cc_mounts.py (hunk 1 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: My bad. Fixed.

File: cloudinit/config/cc_mounts.py (hunk 2 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: I agree with the suggestion and I'll elaborate more on the helper on my next commit.

File: cloudinit/config/cc_mounts.py (hunk 3 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: Agreed as per my last comment. Thanks for the review!

**Comment by powersj** [comment]
Marking WIP till tests are written

File: cloudinit/config/cc_mounts.py [hunk 4]
```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+                   'count=%d' % size]
+
+        try:
+            util.subp(cmd, capture=True)
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, method, e)
+
+            if method == "fallocate":
+                LOG.warning("Will attempt with dd.")
+                create_swap(fname, size, "dd")
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: This behaviour seems a little unexpected to me. I wouldn't expect `create_swap(..., "fallocate")` to do anything other than attempt to use `fallocate` to create the swap.

I think moving this logic up to the calling function would make it clearer what the high-level behaviour is:

```py
    if fstype in ("xfs", "btrfs"):
        create_swap(fname, size, "dd")
    else:
        try:
            create_swap(fname, size, "fallocate")
        except util.ProcessExecutionError as e:
            create_swap(fname, size, "dd")
```

File: cloudinit/config/cc_mounts.py [hunk 5]
```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+
+
+def create_swapfile(fname, size):
+    """Size is in MB."""
+
+    def unlink_f(fname):
+        if os.path.exists(fname):
+            os.unlink(fname)
+
+    def create_swap(fname, size, method):
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: Do these two functions need to be nested?

File: cloudinit/config/cc_mounts.py [hunk 6]
```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+        create_swap(fname, size, "fallocate")
+
+    try:
+        os.chmod(fname, 0o644)
+        util.subp(['mkswap', fname])
+    except util.ProcessExecutionError:
+        unlink_f(fname)
+        raise
+
+    return
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: No need for this `return`.

File: cloudinit/config/cc_mounts.py [hunk 7]
```diff
@@ -245,20 +302,8 @@ def setup_swapfile(fname, size=None, maxsize=None):
...
-                             ('rm -f "$1" && umask 0066 && '
-                              '{ fallocate -l "${2}M" "$1" || '
-                              'dd if=/dev/zero "of=$1" bs=1M "count=$2"; } && '
-                              'mkswap "$1" || { r=$?; rm -f "$1"; exit $r; }'),
-                             'setup_swap', fname, mbsize]])
-
-    except Exception as e:
-        raise IOError("Failed %s: %s" % (msg, e))
+    util.log_time(LOG.debug, msg="Setting up swap file", func=create_swapfile,
+                  args=[fname, int(size / (2 ** 20))])
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: I think having `int(size / (2 ** 20))` in a named variable would make it easier to follow what this operation is.

File: cloudinit/config/cc_mounts.py [hunk 8]
```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
 
+def get_fstype(path):
+    """Return the type of the filesystem that path is on.
+    If path is a mount point itself, then the type of the filesystem mounted
+    on path rather than its parent dir."""
+    return util.get_mount_info(path)[1]
+
+
+def create_swapfile(fname, size):
+    """Size is in MB."""
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: MB or MiB?

File: cloudinit/config/cc_mounts.py [hunk 9]
```diff
@@ -245,20 +302,8 @@ def setup_swapfile(fname, size=None, maxsize=None):
         LOG.debug("Not creating swap: suggested size was 0")
         return
 
-    mbsize = str(int(size / (2 ** 20)))
-    msg = "creating swap file '%s' of %sMB" % (fname, mbsize)
-    try:
-        util.ensure_dir(tdir)
-        util.log_time(LOG.debug, msg, func=util.subp,
-                      args=[['sh', '-c',
-                             ('rm -f "$1" && umask 0066 && '
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: Do we need to do something to replicate this umask call in the new codepaths?

File: tests/unittests/test_handler/test_handler_mounts.py [hunk 10]
```diff
@@ -181,6 +181,18 @@ def device_name_to_device(self, path):
 
         return dev
 
+    def test_swap_integrity(self):
+        '''Ensure that the swap file is correctly created and can
+        swapon succsesfully. Fixing the corener case of:
```
Comment (OddBloke) [inline tests/unittests/test_handler/test_handler_mounts.py]:
```suggestion
        swapon successfully. Fixing the corner case of:
```

File: cloudinit/config/cc_mounts.py [hunk 11]
```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, method, e)
+
+            if method == "fallocate":
+                LOG.warning("Will attempt with dd.")
+                create_swap(fname, size, "dd")
+            else:
+                unlink_f(fname)
+
+    tdir = os.path.dirname(fname)
```
Comment (OddBloke) [inline cloudinit/config/cc_mounts.py]: What does the `t` in `tdir` here stand for?  (My brain interpreted it, incorrectly, as "temp". Is there a more descriptive name we could use?)

File: cloudinit/config/cc_mounts.py [hunk 12]
```diff
@@ -223,6 +223,63 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+    util.ensure_dir(tdir)
+
+    fstype = get_fstype(tdir)
+
+    if fstype in ("xfs", "btrfs"):
+        create_swap(fname, size, "dd")
+    else:
```
Comment (igalic) [inline cloudinit/config/cc_mounts.py]: perhaps before defaulting to fallocate, we should check that it exists?

Comment (OddBloke) [reply cloudinit/config/cc_mounts.py]: So as-written, `create_swap(..., "fallocate")` will fallback to using dd if using fallocate fails (which it obviously would if it's absent).

My suggestion in https://github.com/canonical/cloud-init/pull/70#discussion_r359064993 would make that clearer, I think.

File: cloudinit/config/cc_mounts.py (hunk 5 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: I don't think they specifically need to be nested, but I don't see other usage for create_swap() other than inside this function. It's a very local routine that is called three times. I'm gonna keep this here for now.

File: cloudinit/config/cc_mounts.py (hunk 4 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: Agreed

File: cloudinit/config/cc_mounts.py (hunk 9 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: Actually this umask is wrong, it should be 177 (u=rw,go=) and it's handled on the last try/except of create_swapfile() function.

File: cloudinit/config/cc_mounts.py (hunk 11 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: This name was inherited from the precious version pre-refactoring :-) but yeah, I agree. Perhaps the meaning is temporary, but it's not correct. I'll just change it.

Comment (otubo) [reply cloudinit/config/cc_mounts.py]: Oh I see now. This name is also being used on the function setup_swap(). So the first author of this patch just wanted to keep the same naming scheme, perhaps?

Comment (otubo) [reply cloudinit/config/cc_mounts.py]: I mostly like to keep commits and changes to the smallest related to the topic. But I'm gonna change setup_swap() as well.

File: cloudinit/config/cc_mounts.py (hunk 12 above)
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: Done that.

**Comment by otubo** [comment]
Related bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1772505

**Comment by blackboxsw** [comment]
Eduardo thanks for the contribution here github-side! 

Just one more step so we can correlate old launchpad accounts that have signed the CLA to your github account so that we have accountability for which accounts actually signed the Contribution License Agreement.

Please run ./tools/migrate-lp-user-to-github to allow us to track your ownership of both launchpad and github accounts. Details are in our updated hacking guide https://cloudinit.readthedocs.io/en/latest/topics/hacking.html#do-these-things-once

**Comment by otubo** [comment]
> Eduardo thanks for the contribution here github-side!
> 
> Just one more step so we can correlate old launchpad accounts that have signed the CLA to your github account so that we have accountability for which accounts actually signed the Contribution License Agreement.
> 
> Please run ./tools/migrate-lp-user-to-github to allow us to track your ownership of both launchpad and github accounts. Details are in our updated hacking guide https://cloudinit.readthedocs.io/en/latest/topics/hacking.html#do-these-things-once

Thanks Chad!
I started the process here: 
https://github.com/canonical/cloud-init/pull/135

**Comment by otubo** [comment]
Also, I have to say that I already have a commit on cloud-init on the github repository before I run this tool:
https://github.com/canonical/cloud-init/commit/f1094b1a539044c0193165a41501480de0f8df14
Not sure if there's anything to be done on this commit for some reason :-)

File: cloudinit/config/cc_mounts.py [hunk 13]
```diff
@@ -223,42 +223,86 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+    on path rather than its parent dir."""
+    return util.get_mount_info(path)[1]
+
+
+def create_swapfile(fname, size):
+    """Size is in MiB."""
+
+    errmsg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
+
+    def unlink_f(fname):
```
Comment (smoser) [inline cloudinit/config/cc_mounts.py]: this is util.del_file

File: cloudinit/config/cc_mounts.py [hunk 14]
```diff
@@ -223,42 +223,86 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
     return size
 
 
+def get_fstype(path):
```
Comment (blackboxsw) [inline cloudinit/config/cc_mounts.py]: Do we need this function call that we are just grabbing the 1 index of get_mount_info? We might be able to just move the comment down above to the callsite instead of paying the extra cost of a function call.

File: cloudinit/config/cc_mounts.py [hunk 15]
```diff
@@ -223,42 +223,86 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+    else:
+        try:
+            create_swap(fname, size, "fallocate")
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, "dd", e)
+            LOG.warning("Will attempt with dd.")
+            create_swap(fname, size, "dd")
+
+    try:
+        os.chmod(fname, 0o600)
```
Comment (blackboxsw) [inline cloudinit/config/cc_mounts.py]: os.chmod isn't going to raise a ProcessExecutionError, so we can probably move that up just above  this try scope. It also should probably be util.chmod instead, which accounts for Selinux support too.

File: cloudinit/config/cc_mounts.py [hunk 16]
```diff
@@ -223,42 +223,78 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+        try:
+            util.subp(cmd, capture=True)
+        except util.ProcessExecutionError as e:
+            LOG.warning(errmsg, fname, size, method, e)
+            util.del_file(fname)
+
+    swap_dir = os.path.dirname(fname)
+    util.ensure_dir(swap_dir)
+
+    """Return the type of the filesystem that path is on.
```
[the remark this answers is not in the document]
Comment (otubo) [reply cloudinit/config/cc_mounts.py]: I think I'll just remove it.

**Comment by norrisjeremy** [comment]
This commit broke swap file creation, with the following error:

cc_mounts.py[WARNING]: failed to setup swap: %d format: a number is required, not str

File: cloudinit/config/cc_mounts.py [hunk 17]
```diff
@@ -223,42 +223,75 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+    """Size is in MiB."""
+
+    errmsg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
+
+    def create_swap(fname, size, method):
+        LOG.debug("Creating swapfile in '%s' on fstype '%s' using '%s'",
+                  fname, fstype, method)
+
+        if method == "fallocate":
+            cmd = ['fallocate', '-l', '%dM' % size, fname]
```
Comment (norrisjeremy) [inline cloudinit/config/cc_mounts.py:236]: Size is a string, not a number, so '%s' format specifier should be used.

File: cloudinit/config/cc_mounts.py [hunk 18]
```diff
@@ -223,42 +223,75 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
...
+        elif method == "dd":
+            cmd = ['dd', 'if=/dev/zero', 'of=%s' % fname, 'bs=1M',
+                   'count=%d' % size]
```
Comment (norrisjeremy) [inline cloudinit/config/cc_mounts.py:239]: While I don't use xfs, I suspect this too should be using a '%s' format specifier since size is a string.

File: cloudinit/config/cc_mounts.py [hunk 19]
```diff
@@ -223,42 +223,75 @@ def suggested_swapsize(memsize=None, maxsize=None, fsys=None):
     return size
 
 
+def create_swapfile(fname, size):
+    """Size is in MiB."""
+
+    errmsg = "Failed to create swapfile '%s' of size %dMB via %s: %s"
```
Comment (norrisjeremy) [inline cloudinit/config/cc_mounts.py:229]: This too should likely be a '%s' format modifier instead of '%d' since size is a string.

**Comment by powersj** [comment]
@otubo thoughts on the above comments? 

@norrisjeremy if you have not already can you open a bug: https://bugs.launchpad.net/cloud-init/+filebug

**Comment by norrisjeremy** [comment]
@powersj Please see https://bugs.launchpad.net/cloud-init/+bug/1872836

**Comment by norrisjeremy** [comment]
FYI, I discovered this when attempting to spin up the latest 20.04 beta cloud image and noticed that swapfile creation didn't work in my cloud-config.yml. Hopefully this can be fixed before next weeks release?

**Comment by powersj** [comment]
Fix in https://github.com/canonical/cloud-init/pull/316
